File: Labs/Lab 8.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashSet;
import java.util.Set;

public class CoinChange{
    public int coinChange(int[] coins, int amount){
    if(coins == null || coins.length == 0 || amount < 1) return 0;

    Deque<Integer> queue = new ArrayDeque<Integer>();
    Set<Integer> visited = new HashSet<Integer>();
    queue.addFirst(amount);
    visited.add(amount);
    int level = 0;

    while(!queue.isEmpty()){
      int size = queue.size();

      while(size-- > 0){
        int curr = queue.removeLast();
        if(curr == 0) return level;

        if(curr < 0) continue;

        for(int coin : coins){
          int next = curr - coin;
          if(next >= 0 && !visited.contains(next)){
            queue.addFirst(next);
            visited.add(next);
          }
        }
      }

      level++;
    }

    return -1;
  }
public static void main(String[] args) {

    CoinChange coinChange = new CoinChange();
    // Test Case 1
    int[] coins1 = {1, 2, 5};
    int amount1 = 11;
    System.out.println("Test Case 1:");
    System.out.println("Input: coins = [1,2,5], amount = 11");
    System.out.println("Output: " + coinChange.coinChange(coins1, amount1)); 

    // Test Case 2
    int[] coins2 = {2};
    int amount2 = 3;
    System.out.println("\nTest Case 2:");
    System.out.println("Input: coins = [2], amount = 3");
    System.out.println("Output: " + coinChange.coinChange(coins2, amount2)); 

    // Test Case 3
    int[] coins3 = {1};
    int amount3 = 0;
    System.out.println("\nTest Case 3:");
    System.out.println("Input: coins = [1], amount = 0");
    System.out.println("Output: " + coinChange.coinChange(coins3, amount3)); 
}
}

File: Labs/Lab 8/Lab 8.py
from collections import deque
from typing import List

class CoinChange:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0
        coins.sort(reverse = True)
        seen = set()
        queue = deque()
        queue.append((0,0))
        while queue:
            c, t = queue.popleft()
            if t == amount:
                return c
            for coin in coins:
                if t + coin <= amount and t + coin not in seen:
                    queue.append((c + 1, t + coin))
                    seen.add(t + coin)

        return -1

def main():
    coin_changer = CoinChange()

    coins1 = [1, 2, 5]
    amount1 = 11
    print("Test Case 1:")
    print(f"Input: coins = {coins1}, amount = {amount1}")
    print("Output:", coin_changer.coinChange(coins1, amount1)) 

    coins2 = [2]
    amount2 = 3
    print("\nTest Case 2:")
    print(f"Input: coins = {coins2}, amount = {amount2}")
    print("Output:", coin_changer.coinChange(coins2, amount2))  

    coins3 = [1]
    amount3 = 0
    print("\nTest Case 3:")
    print(f"Input: coins = {coins3}, amount = {amount3}")
    print("Output:", coin_changer.coinChange(coins3, amount3))

if __name__ == "__main__":
    main()
